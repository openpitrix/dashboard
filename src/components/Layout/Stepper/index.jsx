import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withRouter } from 'react-router-dom';
import { observer } from 'mobx-react';
import { translate } from 'react-i18next';
import _ from 'lodash';
import Loading from 'components/Loading';

import { DocLink, Icon } from 'components/Base';

import styles from './index.scss';

@translate()
@observer
class LayoutStepper extends Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    headerCls: PropTypes.string,
    name: PropTypes.string,
    stepOption: PropTypes.shape({
      activeStep: PropTypes.number,
      steps: PropTypes.number,
      prevStep: PropTypes.func,
      disableNextStep: PropTypes.bool,
      isLoading: PropTypes.bool,
      nextStep: PropTypes.func
    }),
    titleCls: PropTypes.string
  };

  renderTopProgress() {
    const { stepOption } = this.props;
    const { steps, activeStep } = stepOption;
    const width = `${activeStep * 100 / steps}%`;
    const className = activeStep > steps ? 'headerStepFinished' : 'headerStepNotFinished';

    const style = {
      width,
      transition: `width 500ms ease`
    };

    return (
      <div
        style={style}
        className={classnames(styles.headerStep, styles[className])}
      />
    );
  }

  renderTopNav() {
    const {
      name, stepOption, t, history
    } = this.props;
    const {
      activeStep, steps, prevStep, goBack
    } = stepOption;
    const funcBack = _.isFunction(goBack) ? goBack : history.goBack;

    if (activeStep > steps) {
      return null;
    }

    let text = '';
    if (activeStep > 1 && !!name) {
      text = t(`STEPPER_HEADER_${name.toUpperCase()}_${activeStep}`);
    }

    return (
      <div className={styles.operate}>
        {activeStep > 1 && (
          <label onClick={prevStep}>
            ‚Üê&nbsp;{t('Back')}&nbsp;
            <span className={styles.operateText}>{text}</span>
          </label>
        )}
        <label className="pull-right" onClick={funcBack}>
          {t('Esc')}&nbsp;
          <Icon name="close" size={20} type="dark" />
        </label>
      </div>
    );
  }

  renderTitle() {
    const {
      name, stepOption, headerCls, titleCls, t
    } = this.props;
    const { activeStep, steps } = stepOption;

    if (activeStep > steps) {
      return null;
    }

    const nameKey = name.toUpperCase();
    const header = t(`STEPPER_NAME_${nameKey}_HEADER`, {
      activeStep,
      steps
    });
    const title = t(`STEPPER_TITLE_${nameKey}_${activeStep}`);

    return (
      <div className={classnames(styles.stepContent)}>
        <div className={classnames(styles.stepName, headerCls)}>{header}</div>
        <div className={classnames(styles.stepExplain, titleCls)}>{title}</div>
      </div>
    );
  }

  renderFooter() {
    const { t, stepOption, name } = this.props;
    const {
      activeStep,
      steps,
      nextStep,
      disableNextStep,
      btnText
    } = stepOption;

    if (activeStep > steps) {
      return null;
    }

    const keyName = `STEPPER_FOOTER_${name.toLocaleUpperCase()}_${activeStep}`;
    const tipText = t(keyName);
    const tipLink = <DocLink name={keyName} />;

    const buttonText = activeStep === steps ? t('Done') : t('Go on');
    const iconName = activeStep === steps ? 'checked-icon' : 'next-icon';

    return (
      <div className={styles.footer}>
        <span className={styles.footerTips}>
          <span className={styles.footerTipsButton}>{t('Tips')}</span>
          {tipText}
          {tipLink}
        </span>
        <button
          className={classnames(styles.button, {
            [styles.buttonActived]: !disableNextStep
          })}
          type="primary"
          onClick={nextStep}
        >
          <Icon className={styles.icon} name={iconName} size={20} />
          <span>{t(btnText || buttonText)}</span>
        </button>
      </div>
    );
  }

  render() {
    const { className, children, stepOption } = this.props;
    return (
      <div className={classnames(styles.layout)}>
        {this.renderTopProgress()}
        {this.renderTopNav()}
        {this.renderTitle()}
        <Loading isLoading={stepOption.isLoading}>
          <div className={classnames(styles.mainContent, className)}>
            {children}
          </div>
        </Loading>
        {this.renderFooter()}
      </div>
    );
  }
}

export default withRouter(LayoutStepper);
