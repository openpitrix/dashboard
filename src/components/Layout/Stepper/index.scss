@import '~scss/vars';

.headerStep {
  position: fixed;
  top: 0;
  left: 0;
  height: 2px;
}

.headerStepNotFinished {
  background-color: $Y75;
}

.headerStepFinished {
  background-color: $G100;
}

.operate {
  position: fixed;
  top: 20px;
  left: 20px;
  right: 20px;
  height: 32px;

  > label {
    display: inline-block;
    font-size: 12px;
    line-height: 32px;
    color: $N300;
    border-radius: 2px;
    cursor: pointer;
    padding: 0 12px;
    // background: $N10;
  }
  :global {
    .icon-close {
      position: static;
      transform: none;
      vertical-align: middle;
      margin-top: -1px;
    }
  }
}
.operateText {
  color: $N65;
}


.footer {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 4;
  height: $footer-height;
  padding: 0 20px;
  display: flex;
  align-items: center;
  background: $N0;
  border-top: 1px solid rgba($N30, 0.5);
}
.footerTips {
  font-size: 12px;
  line-height: 1.67;
  margin: 22px 0;
  display: inline-block;
  color: $N65;
  flex: 1;

  .footerTipsButton {
    border-radius: 1px;
    background-color: $N30;
    padding: 2px 4px;
    color: $N100;
    margin-right: 4px;
    display: inline-block;
  }
}
.button {
  border-radius: 2px;
  color: $N0;
  padding: 4px 12px;
  box-shadow: 0 1px 3px 0 rgba($P400, 0.16), 0 1px 2px 0 rgba($N500, 0.03);
  background-color: $P75;
  border: 0;
  opacity: 0.5;
  cursor: not-allowed;
  > span {
    font-size: 14px;
    line-height: 24px;
    text-align: center;
  }
  &:active {
    border: 0;
  }
}
.buttonActived {
  opacity: 1;
  cursor: pointer;
}


.icon {
  vertical-align: middle;
  margin: 0 4px;
  margin-top: -1px;
  :global {
    .qicon {
      --primary-color: #fff;
    }
  }
}

.stepContent {
  padding: 100px 0 0;
  min-width: 700px;
  text-align: center;

  .stepName {
    margin-bottom: 4px;
    font-size: 14px;
    line-height: 24px;
    color: $N300;
  }

  .stepExplain {
    margin-bottom: 32px;
    font-size: 24px;
    font-weight: 500;
    line-height: 28px;
    letter-spacing: 0;
    color: $N500;
  }
}

.mainContent {
  margin: 0 auto $footer-height + 20px;
  max-width: $create-content-width;
}

