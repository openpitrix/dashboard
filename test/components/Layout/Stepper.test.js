import { Provider } from 'mobx-react';
import { BrowserRouter } from 'react-router-dom';
import { observable } from 'mobx';

import Stepper from 'components/Layout/Stepper';

const renderPage = (store, props) => (
  <Provider rootStore={store}>
    <BrowserRouter>
      <Stepper {...props}>
        <div>test</div>
      </Stepper>
    </BrowserRouter>
  </Provider>
);

describe('Layout/Stepper', () => {
  let store,
    props;

  beforeAll(() => {
    store = observable.object({
      user: {}
    });

    props = {
      name: 'CREATE_APP',
      stepOption: {
        disableNextStep: true,
        nextStep: jest.fn(),
        stepBase: 1,
        activeStep: 1,
        steps: 3
      }
    };
  });

  it('basic render', () => {
    const wrapper = render(renderPage(store, props));

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('basic render activeStep more than the steps', () => {
    props = {
      name: 'CREATE_APP',
      stepOption: {
        activeStep: 4,
        steps: 3
      }
    };

    const wrapper = render(renderPage(store, props));

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
