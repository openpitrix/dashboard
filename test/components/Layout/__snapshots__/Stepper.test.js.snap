// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout/Stepper basic render 1`] = `
<div
  class="stepper"
>
  <div
    class="headerStep headerStepNotFinished"
    style="width:33.333333333333336%;transition:width 500ms ease"
  />
  <div
    class="operate"
  >
    <label
      class="pull-right"
    >
      EscÂ 
      <span
        class="icon icon-close"
        style="width:20px;height:20px"
      >
        <svg
          class="qicon qicon-close qicon-dark"
        >
          <use
            href="#qui-close"
          />
        </svg>
      </span>
    </label>
  </div>
  <div
    class="stepContent"
  >
    <div
      class="stepName"
    >
      STEPPER_NAME_CREATE_APP_HEADER
    </div>
    <div
      class="stepExplain"
    >
      STEPPER_TITLE_CREATE_APP_1
    </div>
  </div>
  <div
    class="mainContent"
  >
    <div>
      test
    </div>
  </div>
  <div
    class="footer"
  >
    <span
      class="footerTips"
    >
      <span
        class="footerTipsButton"
      >
        Tips
      </span>
      STEPPER_FOOTER_CREATE_APP_1
    </span>
    <button
      class="button"
      type="primary"
    >
      <span>
        Go on
      </span>
      <span
        class="icon icon-next-icon icon"
        style="width:20px;height:20px"
      >
        <svg
          class="qicon qicon-next-icon qicon-light"
        >
          <use
            href="#qui-next-icon"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`Layout/Stepper basic render activeStep more than the steps 1`] = `
<div
  class="stepper"
>
  <div
    class="headerStep headerStepFinished"
    style="width:133.33333333333334%;transition:width 500ms ease"
  />
  <div
    class="mainContent"
  >
    <div>
      test
    </div>
  </div>
</div>
`;
